src/main/scala/org/deepdive/ddlog/DeepDiveLogPrettyPrinter.scala

```diff
@@ -93,36 +95,52 @@ object DeepDiveLogPrettyPrinter extends DeepDiveLogHandler {
       case NegationCond(c) => s"[!${printCond(c)}]"
       case CompoundCond(lhs, op, rhs) => {
         op match {
-          case LogicOperator.AND => s"[${printCond(lhs)}, ${printCond(rhs)}]" 
+          case LogicOperator.AND => s"[${printCond(lhs)}, ${printCond(rhs)}]"
           case LogicOperator.OR  => s"[${printCond(lhs)}; ${printCond(rhs)}]"
         }
       }
     }
   }
 
-  def print(cq: ConjunctiveQuery): String = {
-    val printAtom = {a:Atom =>
-      val vars = a.terms map printExpr
-      s"${a.name}(${vars.mkString(", ")})"
+  def printAtom(a: Atom) = {
+    val vars = a.terms map printExpr
+    s"${a.name}(${vars.mkString(", ")})"
+  }
+
+  def printQuantifiedBody(a: QuantifiedBody) : String = {
+    val modifier = a.modifier match {
+      case ExistModifier(negated) => if(negated) "NOT " else "" + "EXISTS"
+      case OuterModifier() => "OPTIONAL"
+      case AllModifier() => "ALL"
+    }
+    val bodyStr = (a.bodies map printBody).mkString(", ")
+    s"${modifier}[${bodyStr}]"
+  }
+
+  def printBody(b: Body) = {
+    b match {
+      case b: Atom => printAtom(b)
+      case b: Cond => printCond(b)
+      case b: QuantifiedBody => printQuantifiedBody(b)
     }
-    val printListAtom = {a:List[Atom] =>
-      s"${(a map printAtom).mkString(",\n    ")}"
+  }
+
+  def print(cq: ConjunctiveQuery): String = {
+
+    def printBodyList(b: List[Body]) = {
+      s"${(b map printBody).mkString(",\n    ")}"
     }
 
-    val conditionList = cq.conditions map {
-      case Some(x) => Some(printCond(x))
-      case None    => None
+    val bodyStr = (cq.bodies map printBodyList).mkString(";\n    ")
+
+    val distinctStr = if (cq.isDistinct) "*" else ""
+    val limitStr = cq.limit match {
+      case Some(s) => s" | ${s}"
+      case None => ""
```

Comment:
Let's try to avoid matching Option type (Some/None) as that usually clutters code.  How about:
```scala
val limitStr = cq.limit map { s => s" | ${s}" } getOrElse("")
```

Or:
```scala
val limitStr = cq.limit map { " | " + _ } getOrElse("")
```